include/alpaka/mem/view/Accessor.hpp

```diff
@@ -0,0 +1,400 @@
+    };
+    struct ReadWriteAccess
+    {
+    };
+
+    template<typename MemoryHandle, typename Elem, typename BufferIdx, std::size_t Dim, typename AccessModes>
+    struct Accessor;
+
+    namespace internal
+    {
+        template<typename T>
+        struct WriteOnlyProxy
+        {
+            WriteOnlyProxy(T& location) : loc(location)
+            {
+            }
+
+            template<typename U>
+            auto& operator=(U&& value)
+            {
+                loc = std::forward<U>(value);
+                return *this;
+            }
+
+        private:
+            T& loc;
+        };
+
+        template<typename Pointer, typename Elem, typename AccessModes>
+        struct AccessReturnTypeImpl;
+
+        template<typename Pointer, typename Elem>
+        struct AccessReturnTypeImpl<Pointer, Elem, ReadAccess>
+        {
+            using type = Elem;
+        };
+
+        template<typename Pointer, typename Elem>
+        struct AccessReturnTypeImpl<Pointer, Elem, WriteAccess>
+        {
+            using type = WriteOnlyProxy<Elem>;
+        };
+
+        template<typename Pointer, typename Elem>
+        struct AccessReturnTypeImpl<Pointer, Elem, ReadWriteAccess>
+        {
+            using type = Elem&;
+        };
+
+        template<typename Pointer, typename Elem, typename HeadAccessMode, typename... TailAccessModes>
+        struct AccessReturnTypeImpl<Pointer, Elem, std::tuple<HeadAccessMode, TailAccessModes...>>
+            : AccessReturnTypeImpl<Pointer, Elem, HeadAccessMode>
+        {
+        };
+    } // namespace internal
+
+    template<typename Pointer, typename Elem, typename AccessModes>
+    using AccessReturnType = typename internal::AccessReturnTypeImpl<Pointer, Elem, AccessModes>::type;
+
+    template<typename Pointer, typename Elem, typename BufferIdx, typename AccessModes>
+    struct Accessor<Pointer, Elem, BufferIdx, 1, AccessModes>
+    {
+        using ReturnType = AccessReturnType<Pointer, Elem, AccessModes>;
+
+        ALPAKA_FN_ACC Accessor(Pointer p_, Vec<DimInt<1>, BufferIdx> extents_) : p(p_), extents(extents_)
+        {
+        }
+
+        template<typename OtherAccessModes>
+        ALPAKA_FN_ACC Accessor(const Accessor<Pointer, Elem, BufferIdx, 1, OtherAccessModes>& other)
+            : p(other.p)
+            , extents(other.extents)
+        {
+        }
+
+        ALPAKA_FN_ACC auto operator[](Vec<DimInt<1>, BufferIdx> i) const -> ReturnType
+        {
+            return (*this)(i[0]);
+        }
+
+        ALPAKA_FN_ACC auto operator[](BufferIdx i) const -> ReturnType
+        {
+            return (*this)(i);
+        }
+
+        ALPAKA_FN_ACC auto operator()(BufferIdx i) const -> ReturnType
+        {
+            return p[i];
+        }
+
+        Pointer p;
+        Vec<DimInt<1>, BufferIdx> extents;
+    };
+
+    template<typename Pointer, typename Elem, typename BufferIdx, typename AccessModes>
+    struct Accessor<Pointer, Elem, BufferIdx, 2, AccessModes>
```

Comment:
Pointer should benamed different to avoid that the user thinks only pointer can be used. Maybe `TStorage`?

Reply:
Well, for this specialization only pointers can be used ;) But the concept Pointer here is big enough to include e.g. smart pointers or SYCL pointers.
The template parameter on the primary template is called `TMemoryHandle`.